### ISSUE TITLE
targetselection.py doesn't work with astropy 1.0.5

### ISSUE THREAD

**Issue by sbailey** [desihub/desitarget#3]
With astropy 1.0.5, targetselection.py dies with:

```
$ python $DESITARGET/bin/targetselection.py /data/legacysurvey/dr1/tractor/000/tractor-0009p187.fits blat.fits
Traceback (most recent call last):
  File "/Users/sbailey/desi/git/desitarget/bin/targetselection.py", line 44, in <module>
    main()
  File "/Users/sbailey/desi/git/desitarget/bin/targetselection.py", line 35, in main
    mask = cut.apply(candidates)
  File "/Users/sbailey/desi/git/desitarget/py/desitarget/internal/npyquery.py", line 145, in apply
    result[s] = v.visit(self)
  File "/Users/sbailey/desi/git/desitarget/py/desitarget/internal/npyquery.py", line 369, in visit
    return self.m[t](node)
  File "/Users/sbailey/desi/git/desitarget/py/desitarget/internal/npyquery.py", line 415, in visit_expr
    ops = [self.visit(a) for a in node.operands]
  File "/Users/sbailey/desi/git/desitarget/py/desitarget/internal/npyquery.py", line 369, in visit
    return self.m[t](node)
  File "/Users/sbailey/desi/git/desitarget/py/desitarget/internal/npyquery.py", line 415, in visit_expr
    ops = [self.visit(a) for a in node.operands]
  File "/Users/sbailey/desi/git/desitarget/py/desitarget/internal/npyquery.py", line 369, in visit
    return self.m[t](node)
  File "/Users/sbailey/desi/git/desitarget/py/desitarget/internal/npyquery.py", line 409, in visit_column
    return node.apply(self.array, self.s)
  File "/Users/sbailey/desi/git/desitarget/py/desitarget/internal/npyquery.py", line 243, in apply
    return array[self.name][s]
  File "/Users/sbailey/anaconda/envs/numpy1.9/lib/python2.7/site-packages/astropy/io/fits/fitsrec.py", line 480, in __getitem__
    return self.field(key)
  File "/Users/sbailey/anaconda/envs/numpy1.9/lib/python2.7/site-packages/astropy/io/fits/fitsrec.py", line 620, in field
    field = _get_recarray_field(base, name)
  File "/Users/sbailey/anaconda/envs/numpy1.9/lib/python2.7/site-packages/astropy/io/fits/fitsrec.py", line 1138, in _get_recarray_field
    field = np.recarray.field(array, key)
  File "/Users/sbailey/anaconda/envs/numpy1.9/lib/python2.7/site-packages/numpy/core/records.py", line 542, in field
    res = fielddict[attr][:2]
KeyError: 'brick_primary'
```
This works with astropy 1.0.1.  I'll dig deeper and submit an astropy bug report and/or see if there is a workaround for desitarget.

**Comment by sbailey** [comment]
The problematic piece of code is in desitarget.io.read_tractor:
```
# convert all fields to upper case. 
# We expect normalized upper case field names.
data.dtype.names = [name.upper() for name in data.dtype.names]
```
This may be related to [astropy issue 4105](https://github.com/astropy/astropy/issues/4105) where they broke the treatment of `upper` vs. `lower` keywords starting in 1.0.2.  In general it is dangerous to muck with the underlying data structure anyway, especially when read in via `astropy.io.fits`.  Although it returns a subclass of `numpy.ndarray`, they changed the underlying internal data structure such that it doesn't actually work as a `numpy.ndarray` in all cases.  e.g. see [closed astropy issue 4233](https://github.com/astropy/astropy/issues/4233) where they admit that they broke it and say they won't fix it.

I'll submit a pull request that works around this.  If you don't try to be fancy, `FITS_rec` will support both upper and lowercase field names regardless of what was in the file.

Note: If we switch to fitsio (which returns a real numpy.ndarray without trying to be fancy), `fitsio.read()` has `upper` and `lower` keywords for exactly this purpose -- to normalize the capitalization of the column names regardless of what was in the file.

**Comment by rainwoodman** [comment]
Indeed. The renaming trick was copied from https://github.com/astropy/astropy/blob/master/astropy/io/fits/convenience.py#L202  that becomes broken.

It is a *shame* there is not even a way to construct normalized column names in pyfits. 

My concern was that if the FITS_rec object is converted to an array, the case of the column names are totally unpredictable. That may cause hideous bugs later on. 

But still better than dying with newer astropy.

**Comment by rainwoodman** [comment]
There really is not reason why FITS_rec shall be a subclass of recarray. I've filed an issue report to them. 

Once they break free from recarray it will be obvious they do need to provide a member function to convert FITS_rec to ndarray (with selected columns) and that's where the case normalization shall be. 

Finger crossed.

**Comment by weaverba137** [comment]
Maybe I'm dumb, but if you have the file, why can't you use the column names that are in the file?

**Comment by rainwoodman** [comment]
Well. It is possible, but messy to predict the mapping between a column defined in ts cuts and a unnormalized column name in the fits file.

**Comment by weaverba137** [comment]
In addition to having the file, we also have a data model for every file.  So there should be no need to predict the column names.  They are in the data model!

What is the big deal with normalized column names all of a sudden?  Seriously, I don't get it.

**Comment by rainwoodman** [comment]
The TS cuts on the wiki are in upper case, but the tractor files read by
astropy are in lower case.

It is an anomaly that has to be addressed at some level.

The easiest resolution is to normalize the names immediately after data is
open. Of course relying on FITSrec is another solution. (Which we currently
use)

No big deal.

**Comment by weaverba137** [comment]
Am I reading this correctly?  This is only a problem because you can't change the *wiki*?

**Comment by sbailey** [comment]
Part of the underlying problem is that FITS is a case-insensitive format.  So the input data could be written with upper- or lowercase column names and both would be completely valid representations of the exact same data.  Code is supposed to be able to treat them identically.  If we add a constraint on the capitalization of the column names in the file itself, then we are adding a non-standard constraint on the FITS file format.

However, I don't think this is an actual problem with the current code.  the FITS_rec objects returned by astropy.io.fits already handle both upper and lower case.  e.g. blat['x'] and blat['X'] will both refer to the same column whether or not the underlying file used upper or lower case to describe the column name in the header.  This is part of why astropy.io.fits doesn't use a genuine ndarray -- to provide more formal support of the underlying fits standard agnostic capitalization.  The alternate python fits library, fitsio, returns genuine numpy.ndarray that has to be either upper or lower case, but it also provides keywords for picking which to use when reading in the file to normalize the case.

One might be concerned about casting a FITS_rec object to an ndarray and what case the columns will have, but unfortunately you should never do this anyway since FITS_rec will produce a broken ndarray that can't be used with numpy functions like hstack anyway.  The astropy developers apparently don't plan to fix this, see closed [astropy issue 4233](https://github.com/astropy/astropy/issues/4233).

**Comment by sbailey** [comment]
This was fixed in PR #4 since FITS_rec already supports UPPER or lower case column names in the code transparently of which was used by the input file.  So I'm closing this ticket (but keep commenting on this thread if you'd like).
